sw/device/lib/crypto/drivers/entropy_kat.c

```diff
@@ -0,0 +1,169 @@
+
+  // The following bit indexes are defined in
+  // https://docs.opentitan.org/hw/ip/csrng/doc/#working-state-values
+  state->instantiated = bitfield_bit32_read(flags, /*bit_index=*/0u);
+  state->fips_compliance = bitfield_bit32_read(flags, /*bit_index=*/1u);
+}
+
+/**
+ * Checks the CSRNG internal state against `expected` values.
+ *
+ * @param csrng A CSRNG handle.
+ * @param expected Expected CSRNG internal state.
+ */
+static status_t check_internal_state(
+    const entropy_csrng_internal_state_t *expected) {
+  entropy_csrng_internal_state_t got = {0};
+  entropy_csrng_internal_state_get(kCsrngInternalStateIdSw, &got);
+  if (memcmp(&got, expected, sizeof(entropy_csrng_internal_state_t)) == 0) {
+    return OK_STATUS();
+  }
+  return INTERNAL();
+}
+
+status_t entropy_csrng_kat(void) {
+  TRY(entropy_csrng_uninstantiate());
+
+  const entropy_seed_material_t kEntropyInput = {
+      .data = {0x73bec010, 0x9262474c, 0x16a30f76, 0x531b51de, 0x2ee494e5,
+               0xdfec9db3, 0xcb7a879d, 0x5600419c, 0xca79b0b0, 0xdda33b5c,
+               0xa468649e, 0xdf5d73fa},
+      .len = 12,
+  };
+  TRY(entropy_csrng_instantiate(
+      /*disable_trng_input=*/kHardenedBoolTrue, &kEntropyInput));
+
+  const entropy_csrng_internal_state_t kExpectedStateInstantiate = {
+      .reseed_counter = 1,
+      .v = {0x06b8f59e, 0x43c0b2c2, 0x21052502, 0x217b5214},
+      .key = {0x941709fd, 0xd8a25860, 0x861aecf3, 0x98a701a1, 0x0eb2c33b,
+              0x74c08fad, 0x632d5227, 0x8c52f901},
+      .instantiated = true,
+      .fips_compliance = false,
+  };
+  if (!status_ok(check_internal_state(&kExpectedStateInstantiate))) {
+    return INTERNAL();
+  }
+
+  enum {
+    kExpectedOutputLen = 16,
+  };
+
+  uint32_t got[kExpectedOutputLen];
+  TRY(entropy_csrng_generate(/*seed_material=*/NULL, got, kExpectedOutputLen));
+  TRY(entropy_csrng_generate(/*seed_material=*/NULL, got, kExpectedOutputLen));
+
+  const entropy_csrng_internal_state_t kExpectedStateGenerate = {
+      .reseed_counter = 3,
+      .v = {0xe73e3392, 0x7d2e92b1, 0x1a0bac9d, 0x53c78ac6},
+
+      .key = {0x66d1b85a, 0xc19d4dfd, 0x053b73e3, 0xe9dc0f90, 0x3f015bc8,
+              0x4436e5fd, 0x1cccc697, 0x1a1c6e5f},
+      .instantiated = true,
+      .fips_compliance = false,
+  };
+  if (!status_ok(check_internal_state(&kExpectedStateGenerate))) {
+    return INTERNAL();
+  }
+
+  // TODO(#13342): csrng does not provide a linear output order. For example,
+  // note the test vector output word order: 12,13,14,15 8,9,10,11 4,5,6,7
+  // 0,1,2,3.
+  const uint32_t kExpectedOutput[kExpectedOutputLen] = {
+      0xe48bb8cb, 0x1012c84c, 0x5af8a7f1, 0xd1c07cd9, 0xdf82ab22, 0x771c619b,
+      0xd40fccb1, 0x87189e99, 0x510494b3, 0x64f7ac0c, 0x2581f391, 0x80b1dc2f,
+      0x793e01c5, 0x87b107ae, 0xdb17514c, 0xa43c41b7,
+  };
+  if (!memcmp(got, kExpectedOutput, sizeof(kExpectedOutput))) {
+    return OK_STATUS();
+  }
+
+  return INTERNAL();
```

Comment:
One other bit of commentary:

All of the error macros (`INTERNAL()`, `INVALID_ARGUMENT()`, etc) encode the first three letters of the source file and the line number where the macro was invoked.  This particular error code can be thought of as:
```
return {error: true, module: "ENT", arg: 168, code: kInternal};
```

You may want to set the module_id at the top of every C file to control the name.  For, example, it _might_ be better for this file to:

```
DECLARE_MODULE_ID('K', 'A', 'T');
```
Additionally, if you want the `arg` to be something other than the line number in the source file, you can use any value in the range `0..2047` as an optional argument to the error macros:
```
// Think of this as {error: true, module: "KAT", arg: 13, code: kInternal}
return INTERNAL(13);  // An extremely unlucky internal error.
```

For the `OK_STATUS()` macro, you may use any positive value in the range `0..INT_MAX`, hardening arguments notwithstanding.
```
// Think of this as {error: false, value: 777}
return OK_STATUS(777); // Vegas baby!
```

Reply:
Captured in https://github.com/lowRISC/opentitan/issues/14549 to continue the discussion there.